src/backend/optimizer/plan/createplan.c

```diff
@@ -9063,4 +9064,148 @@ push_locus_down_after_elide_motion(Plan* plan)
+				subplans = ((Append *) (plan))->appendplans;
+				foreach(cell, subplans)
+				{
+					replace_entry_locus_with_singleqe(lfirst(cell));
+				}
+				break;
+			}
+			case T_SubqueryScan:
+				plan = ((SubqueryScan *)(plan))->subplan;
+				break;
+			case T_NestLoop:
+			case T_MergeJoin:
+			case T_HashJoin:
+				replace_entry_locus_with_singleqe(plan->righttree);
+				/* FALLTHROUGH */
+			default:
+				plan = plan->lefttree;
+		}
+	}
+}
+
+/*
+ * Check whether we can safely offload root slice on QD to a QE.
+ */
+static bool
+safe_to_offload_entry_to_qe_rte_walker(List *rtes)
+{
+	ListCell *lc;
+	foreach(lc, rtes)
+	{
+		RangeTblEntry *rte = lfirst_node(RangeTblEntry, lc);
+		if (rte->rtekind == RTE_RELATION)
+		{
+			// check if any partition of a partitioned table is a coordinator-only external/foreign table
+			if (rte->relkind == RELKIND_PARTITIONED_TABLE)
+			{
+				Relation rel;
+				PartitionDesc desc;
+
+				rel = relation_open(rte->relid, NoLock);
+				desc = RelationGetPartitionDesc(rel, true);
+				relation_close(rel, NoLock);
+				for (int i = 0; i < desc->nparts; i++)
+				{
+					if (GpPolicyIsEntry(GpPolicyFetch(desc->oids[i])))
+						return false;
+				}
+				return true;
+			}
+			else
+				return !GpPolicyIsEntry(GpPolicyFetch(rte->relid));
+		}
+		else if (rte->rtekind == RTE_SUBQUERY)
+		{
+			if (!safe_to_offload_entry_to_qe_rte_walker(rte->subquery->rtable))
+				return false;
+		}
+	}
+	return true;
+}
+
+/*
+ * Check if the root slice contains Sort, Join or Aggregate.
+ */
+static bool
+should_offload_entry_to_qe_plan_walker(Plan *plan)
+{
+	while (plan && plan->locustype == CdbLocusType_Entry)
+	{
+		switch (nodeTag(plan))
+		{
+			case T_Motion:
+				return false;
+			/* sort */
+			case T_Sort:
+			/* aggregates*/
+			case T_Agg:
+			case T_WindowAgg:
+			/* join */
+			case T_Join:
+			case T_MergeJoin:
+			case T_HashJoin:
+			case T_NestLoop:
+				return true;
+			default:
+				plan = plan->lefttree;
+		}
+	}
+	return false;
+}
+
+Plan *
+offload_entry_to_qe(PlannerInfo *root, Plan *plan)
+{
+	if (enable_offload_entry_to_qe &&
+		root->parse->commandType == CMD_SELECT &&
+		root->numMotions > 0 &&
+		should_offload_entry_to_qe_plan_walker(plan) &&
+		safe_to_offload_entry_to_qe_rte_walker(root->parse->rtable) &&
+		!contain_volatile_functions((Node *) root->parse))
+	{
+		CdbPathLocus entrylocus;
+		PlanSlice *sendSlice;
+		{
+			sendSlice = (PlanSlice *) palloc0(sizeof(PlanSlice));
+			sendSlice->gangType = GANGTYPE_SINGLETON_READER;
+			sendSlice->numsegments = 1;
+			sendSlice->sliceIndex = -1;
+			sendSlice->parallel_workers = plan->parallel;
+			sendSlice->segindex = gp_session_id % getgpsegmentCount();
+		}
+
+		replace_entry_locus_with_singleqe(plan);
+
+		plan = (Plan *) make_union_motion(plan);
+		((Motion *) plan)->senderSliceInfo = sendSlice;
+
+		plan->locustype = CdbLocusType_Entry;
+		CdbPathLocus_MakeEntry(&entrylocus);
+		plan->flow = cdbpathtoplan_create_flow(root, entrylocus);
```

Comment:
need to free origin flow ?

Reply:
how to free anything in CBDB? should I call `free`? I thought `MemoryContext` already take care of it 🤔

Reply:
but free origin flow as soon as possible.

Reply:
```suggestion
		pfree(plan->flow);
		plan->flow = cdbpathtoplan_create_flow(root, entrylocus);
```
would this resolve your concern?